sklearn_numba_dpex/common/matmul.py

```diff
@@ -0,0 +1,277 @@
+import math
+from functools import lru_cache
+
+import numba_dpex as dpex
+import numpy as np
+
+from sklearn_numba_dpex.common._utils import _check_max_work_group_size
+
+zero_idx = np.int64(0)
+
+
+@lru_cache
+def make_matmul_2d_kernel(
+    X_n_rows,
+    Y_t_n_rows,
+    n_cols,
+    work_group_size,
+    sub_group_size,
+    dtype,
+    device,
+    multiply_fn=None,
+    out_fused_elementwise_fn=None,
+):
+    """Returns a matmul kernel.
+
+    This kernel is about doing a matrix product between `X` and `Y`, but `Y` will be
+    repesented with its transposed `Y_t`, with `X` and `Y_t` being both expected to be
+    C-contiguous, and the (equivalent) matrix product being really computed being
+    between `X` and the transpose of `Y_t`. Hence the choice of using the variable name
+    `Y_t_n_rows` rather than `Y_n_cols`.
+
+    If `out_fused_elementwise_fn` is not None, it will be applied element-wise once to
+    each element of the output array right before it is returned. This function is
+    compiled and fused into the first kernel as a device function with the help of
+    `dpex.func`. This comes with limitations as explained in:
+
+    https://intelpython.github.io/numba-dpex/latest/user_guides/kernel_programming_guide/device-functions.html # noqa
+
+    It is expected to take one scalar argument and return one scalar value.
+
+    Likewise, if `multiply_fn` is not None it will be used in place of the scalar
+    multiply operator . It is expected to take two scalar argument and return one scalar
+    value.
```

Comment:
```suggestion
    Likewise, if `multiply_fn` is not None it will be used in place of the scalar
    multiply operator. It is expected to take two scalar arguments and return one
    scalar value.
```